### DISCUSSION TITLE
MPSC channel assertion error

### DISCUSSION THREAD

**Discussion by TomPridham** [tokio-rs/tokio#4276] [Q&A] [answered]
i am interfacing with the Windows Events api via the windows-rs crate. it requires that i pass a callback that will get called anytime a new event is posted. i am running into an odd intermittent error on the following line: https://docs.rs/tokio/1.14.0/src/tokio/sync/mpsc/chan.rs.html#243. in our real app it doesn't crash the entire process, but in the minimal repro it does. not sure what the difference there is. but the error we get is `thread 'main' panicked at 'assertion failed: self.inner.semaphore.is_idle()', C:\Users\_\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.14.0\src\sync\mpsc\chan.rs:252:13`
the repro is here: https://github.com/TomPridham/windows-events-rust
if anyone has some insight into what i am doing wrong here, it would be much appreciated

**Comment by Darksonn** [comment, marked as answer]
Does that callback get called more than once? Because your use of `Box::from_raw` is deallocating the `Box` after the first time its called.

**Reply by TomPridham** [reply]
oh, duh. somehow i missed that it took ownership and deallocated it. changing that line to this gets things working correctly. thanks so much!
```
-    let sender = unsafe { Box::from_raw(p_context as *mut UnboundedSender<String>) };
+    let sender = unsafe {
+        (p_context as *mut UnboundedSender<String>)
+            .as_ref()
+            .unwrap()
+    };
